### ISSUE TITLE
`berkeley-schema-fy24`: id slot_usage resulting in badly minted IDs

### ISSUE THREAD

**Issue by kheal** [microbiomedata/nmdc-schema#2122]
In `berkeley-schema-fy24`, some of the syntax declarations on ids of classes result in extra parentheses during minting on the berkeley-schema nmdc runtime.  This results in badly minted ids for instances of new classes.

TLDR.  
Actions to take to fix this bug:
- [x] Shorten id typecodes to adhere to nmdc standards and update CONTRIBUTING guide accordingly (in berkeley schema repo)
- [x] Remove extraneous parentheses for ids on definition of `MassSpectrometryConfiguration` and  `ChromatographyConfiguration` (in berkeley schema repo)
- microbiomedata/nmdc-runtime#592

For example:
```
  ChromatographyConfiguration:
    is_a: Configuration
    class_uri: nmdc:ChromatographyConfiguration
    slot_usage:
      id:
        structured_pattern:
          syntax: "{id_nmdc_prefix}:(chrocon)-{id_shoulder}-{id_blade}$"
          interpolated: true
```
Example minted ID of `ChromatographyConfiguration` = nmdc:(chrocon)-14-bwmbj819.  
That can be fixed easily with a schema change by removing the () around chrocon, as classes without this result in well minted IDs (example `Instrument` class mints to nmdc:inst-14-3p938v35 for example)

**However** instances of `DataGeneration`:`NucleotideSequencing` are less easily fixed
```
  NucleotideSequencing:
    class_uri: nmdc:NucleotideSequencing
    is_a: DataGeneration
    slot_usage:
      id:
        structured_pattern:
          syntax: "{id_nmdc_prefix}:(dgns|omprc)-{id_shoulder}-{id_blade}$"
          interpolated: true
```
Example minted ID of `NucleotideSequencing` = nmdc:(dgms|omprc)-14-2204gm61. 🙀 
My guess is that the minter doesn't know which to choose of those options (dgms|omprc).
My understanding is that we will need to keep the flexibility on the id prefixes for referential integrity, so I do not have a solution forward to fixing this example.

I am not sure _where_ to fix this - on the schema or in the minter or some combination of both?

@brynnz22, @eecavanna, @aclum, @dwinston, @turbomam - I'm actually not sure who all needs to be in on this conversation since it's a schema:minter intersection issue.

**Comment by eecavanna** [comment]
Thanks for reporting this, @kheal, for prefixing the issue name with `berkeley-schema-fy24:`, and including the `(dgns|omprc)` example.

You have already CC'ed @dwinston. I'll also CC @PeopleMakeCulture here so she is aware of the issue.

I don't know how the minter gets the value to include in the ID it mints. I'm going to add this to "[Update the Runtime to work with the Berkeley schema](https://github.com/microbiomedata/nmdc-runtime/issues/577)" meta issue, although—like you—I don't know where this will be fixed (in terms of the schema versus the runtime, or both).

**Comment by turbomam** [comment]
In nmdc-schema

```shell
grep -c -R 'syntax: "{id_nmdc_prefix}:' src/schema/*
```

> src/schema/annotation.yaml:1
> src/schema/basic_slots.yaml:0
> src/schema/core.yaml:4
> src/schema/external_identifiers.yaml:0
> src/schema/mixs.yaml:0
> src/schema/nmdc.yaml:37
> src/schema/portal/emsl.yaml:0
> src/schema/portal/jgi_metagenomics.yaml:0
> src/schema/portal/.gitkeep:0
> src/schema/portal/mixs_inspired.yaml:0
> src/schema/portal/jgi_metatranscriptomics.yaml:0
> src/schema/portal/sample_id.yaml:0
> src/schema/prov.yaml:3
> src/schema/workflow_execution_activity.yaml:16


```shell
grep -c -R "syntax: '{id_nmdc_prefix}:" src/schema/*
```

> src/schema/annotation.yaml:0
> src/schema/basic_slots.yaml:0
> src/schema/core.yaml:0
> src/schema/external_identifiers.yaml:0
> src/schema/mixs.yaml:0
> src/schema/nmdc.yaml:1
> src/schema/portal/emsl.yaml:0
> src/schema/portal/jgi_metagenomics.yaml:0
> src/schema/portal/.gitkeep:0
> src/schema/portal/mixs_inspired.yaml:0
> src/schema/portal/jgi_metatranscriptomics.yaml:0
> src/schema/portal/sample_id.yaml:0
> src/schema/prov.yaml:0
> src/schema/workflow_execution_activity.yaml:0

**Comment by turbomam** [comment]
similar pattern in `berkeley-schema-fy24`

**Comment by kheal** [comment]
The parentheses wrapped, pipe-separated list of typecodes are appropriate for the structured pattern checks that linkML does, but the minter seems to use the id slot to make ids.  As written, how would the minter know which of the type codes to use in  `{id_nmdc_prefix}:(dgns|omprc)-{id_shoulder}-{id_blade}$` when generating the IDs.  Therefore, I think the issue is only when the syntax `{id_nmdc_prefix}:(` is declared on the `id.structured_pattern.syntax` slot.

**Comment by turbomam** [comment]
@kheal can you please show some minting API calls that result in useful vs mangled `id` minting?

**Comment by kheal** [comment]
@brynnz22  was doing the actual minting while I was schema hunting to track down the source.

Brynn, could you add an example API call for 
1) `Instrument` [useful] 
2) `MassSpectrometryConfiguration` [mangled, but easily fixable with schema mods]
3) `MassSpectrometry` [mangled, and not easily fixable with schema mods alone]

**Comment by turbomam** [comment]
I think I see how to do it. I'll authenticate and try again.

I guess we should ask for the example payload to be updated!

```
{
  "schema_class": {
    "id": "nmdc:AnalyticalSample"
  },
  "how_many": 1
}
```

**Comment by turbomam** [comment]
Well I'm certainly in favor of removing the un-necessary parentheses wrappers from both schemas right away if you want.

This is the closest thing I have found to discovering them so far:

```shell
grep -C 3 -R 'syntax: "{id_nmdc_prefix}:(' src/schema/*
```

**Comment by turbomam** [comment]
Well I'm certainly in favor of removing the un-necessary parentheses wrappers from both schemas right away if you want.

This is the closest thing I have found to discovering them so far:

```shell
grep -C 3 -R 'syntax: "{id_nmdc_prefix}:(' src/schema/*"
```

**Comment by turbomam** [comment]
This works in PyCHarm's find in files:

```
id\:\s+structured_pattern\:\s+syntax: "\{id_nmdc_prefix\}\:\(
```

**Comment by turbomam** [comment]
@aclum do you think `NucleotideSequencing` and `MassSpectrometry` currently need (or will always need) those dual-typecode patterns?

**Comment by kheal** [comment]
> **aside**: do we have any better specifications on how many characters can be in a typecode?
> 
> * [We do not currently have any criteria for good typecodes, other than they are ideally between 3 and 6 characters in length.](https://github.com/microbiomedata/nmdc-schema/blob/c15b439576d9c8c72cb431133c9fd2a21cb37611/src/docs/prefixes_curies_ids_mappings_etc.md?plain=1#L187)
> * [^(?nmdc):(?[a-z]{1,6})-(?[0-9][a-z]{0,6}[0-9])-(?[A-Za-z0-9]+)(?(.[A-Za-z0-9]+)*)(?_[A-Za-z0-9_.-]+)?$](https://github.com/microbiomedata/nmdc-schema/blob/c15b439576d9c8c72cb431133c9fd2a21cb37611/src/docs/identifiers.md?plain=1#L148)
> 
> 6 characters appears as the max in a few suggestions. These are meant to be very short mnemonics. For example, `storpro` could follow the example of `poolp`, ie `p`p not `pro`. `chrocon` could be `chrcon`
> 
> nmdc-schema typecodes
> berkeley-schema-fy24 typecodes
> **aside**: I would like us to pick a single minting authority code for all of our example data, like 00 or 99.

I have no problem shortening the typecodes for `chrocon`  and `storpro` if 6 characters is preferred, but we need to do it ASAP because we are in the process of generating metadata for at least one of those classes.  If we do decide to go that route I would also ask that standard to be added to the CONTRIBUTING.md under "Modeling Best Practices" for folks like me who are not fluent in ontology nuances.

**Comment by aclum** [comment]
This is from 
https://microbiomedata.github.io/nmdc-schema/identifiers/
<typecode>: An alphabetical code with a 1:1 correspondence to a class from the NMDC schema. Answers the question "of what class is the data record that bears this id"? Must consist of 1 to 6 lower case letters, although a minimum of 3 letters is suggested. **The type code portion of an NMDC id must match the regular expression [a-z]{1,6}.**

The original motivation, decided at the berkeley in person meeting,  behind supporting both (dgms|omprc) and (dgns|omprc) was to support the legacy typecodes so we didn't have to change ids when typecode changed. Doing otherwise sets a bad precedent in which our identifiers are potentially not persistent.  I still that decision which means we'll need to support the legacy typecodes going forward.

**Comment by kheal** [comment]
Thanks @aclum.  I'll summarize what I think of are actionable steps going forward to close this issue. 

- [ ] Shorten id typecodes to adhere to nmdc standards and update CONTRIBUTING guide accordingly (in berkeley schema repo)
- [ ] Remove extraneous parentheses for ids on definition of `MassSpectrometryConfiguration` and  `ChromatographyConfiguration` (in berkeley schema repo)
- [ ] Adjust minting when id has multiple typecodes (like "(dgms|omprc)") to mint IDs with the _first_ typecode (in nmdc runtime)

I will fix the first two so we can review and merge during the metadata meeting tomorrow to unblock refactoring work in progress.  I'll convert the third into a separate issue in the nmdc runtime repo and let @eecavanna or @dwinston address that issue.

**Comment by brynnz22** [comment]
> @brynnz22 was doing the actual minting while I was schema hunting to track down the source.
> 
> Brynn, could you add an example API call for
> 
> 1. `Instrument` [useful]
> 2. `MassSpectrometryConfiguration` [mangled, but easily fixable with schema mods]
> 3. `MassSpectrometry` [mangled, and not easily fixable with schema mods alone]

@turbomam if you want to test and get the API calls you can go the Berkeley runtime and mint these ids to see if you haven't already done so: https://api-berkeley.microbiomedata.org/docs#/minter/mint_ids_pids_mint_post

**Comment by eecavanna** [comment]
FYI: I created a PR in `nmdc-runtime` with what I think is a fix for the third item. PR: https://github.com/microbiomedata/nmdc-runtime/pull/597/. I've requested feedback (there) from the Polyneme team members.

**Comment by turbomam** [comment]
> @turbomam if you want to test and get the API calls you can go the Berkeley runtime and mint these ids to see if you haven't already done so: https://api-berkeley.microbiomedata.org/docs#/minter/mint_ids_pids_mint_post

Thanks @brynnz22 !

see also 
- https://github.com/microbiomedata/nmdc-schema/issues/2122#issuecomment-2231393912

**Comment by eecavanna** [comment]
As mentioned [here](https://github.com/microbiomedata/berkeley-schema-fy24/pull/225#issuecomment-2234935926) (verbatim): The [Minter in the Berkeley environment](https://api-berkeley.microbiomedata.org/docs#/minter/mint_ids_pids_mint_post) has been [updated](https://github.com/microbiomedata/nmdc-runtime/pull/597) to (a) tolerate one layer of extraneous parentheses and (b) to use the first typecode — `foo` — in a `(foo|bar|baz)` list.
